src/simulators/statevector/qvintrin_avx.hpp

```diff
@@ -0,0 +1,955 @@
+      _apply_matrix_double_avx_q0q1(real, img, dmat, inds, qregs);
+    };
+
+    _apply_lambda(data_size, 1, lambda, qregs, omp_threads, mat);
+
+  } else if (qregs[0] < 2) {
+    auto lambda = [&](const areg_t<(1 << N)>& inds, const QV::cvector_t<double>& dmat) -> void {
+      _apply_matrix_double_avx_qL(real, img, dmat, inds, qregs);
+    };
+
+    _apply_lambda(data_size, 2, lambda, qregs, omp_threads, mat);
+
+  } else {
+    auto lambda = [&](const areg_t<(1 << N)>& inds, const QV::cvector_t<double>& dmat) -> void {
+      _apply_matrix_double_avx(real, img, dmat, inds, qregs);
+    };
+
+    _apply_lambda(data_size, 4, lambda, qregs, omp_threads, mat);
+
+  }
+  return Avx::Applied;
+}
+
+
+template<typename FloatType>
+typename std::enable_if<std::is_same<FloatType, double>::value, bool>::type
+is_simd_applicable(uint64_t data_size){
+  if (data_size <= 4)
+    return false;
+  return true;
+}
+
+template<typename FloatType>
+typename std::enable_if<std::is_same<FloatType, float>::value, bool>::type
+is_simd_applicable(uint64_t data_size){
+  if (data_size <= 8)
+    return false;
+  return true;
+}
+
+template<typename FloatType, size_t N>
+inline Avx apply_matrix_avx(
+    std::complex<FloatType>* data,
+    uint64_t data_size,
+    const areg_t<N>& qregs,
+    const cvector_t<FloatType>& mat,
+    uint_t omp_threads
+){
+
+  if (!is_simd_applicable<FloatType>(data_size))
+    return Avx::NotApplied;
+
+  auto transpose = [](const cvector_t<FloatType>& matrix) -> cvector_t<FloatType> {
+      cvector_t<FloatType> transposed(matrix.size());
+      // We deal with MxM matrices, so let's take rows for example
+      auto rows = sqrt(matrix.size());
+      for(size_t i = 0; i < rows; ++i){
+          for(size_t j = 0; j < rows; ++j){
+              transposed[ i * rows + j] = matrix[ j * rows + i ];
+          }
+      }
+      return transposed;
+  };
+
+  auto transposed_mat = transpose(mat);
+  auto ordered_qregs = qregs;
+  reorder(ordered_qregs, transposed_mat);
+  return _apply_avx_kernel<N>(ordered_qregs, data, data_size, transposed_mat, omp_threads);
+}
+
+template<typename FloatType>
+inline Avx apply_matrix_avx(
+  std::complex<FloatType>* qv_data,
+  uint64_t data_size,
+  const reg_t& qregs,
+  const cvector_t<FloatType>& mat,
+  uint_t omp_threads
+){
+
+  switch (qregs.size()) {
+  case 1:
+    return apply_matrix_avx(qv_data, data_size, to_array<1>(qregs), mat, omp_threads);
+  case 2:
+    return apply_matrix_avx(qv_data, data_size, to_array<2>(qregs), mat, omp_threads);
+  case 3:
+    return apply_matrix_avx(qv_data, data_size, to_array<3>(qregs), mat, omp_threads);
+  case 4:
+    return apply_matrix_avx(qv_data, data_size, to_array<4>(qregs), mat, omp_threads);
+  case 5:
+    return apply_matrix_avx(qv_data, data_size, to_array<5>(qregs), mat, omp_threads);
+  case 6:
+    return apply_matrix_avx(qv_data, data_size, to_array<6>(qregs), mat, omp_threads);
+  default:
```

Comment:
We we use block-matrix multiplication to apply larger SIMD matrices using SIMD blocks?